src/lib/snark_work_lib/partitioned.ml

```diff
@@ -0,0 +1,86 @@
+open Core_kernel
+
+(* A `Pairing.t` identifies a single work in Work_selector's perspective *)
+module Pairing = struct
+  module UUID = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        (* this identifies a One_or_two work from Work_selector's perspective *)
+        type t = Pairing_UUID of int
+        [@@deriving compare, hash, sexp, yojson, equal]
+
+        let to_latest = Fn.id
+      end
+    end]
+  end
+
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      (* Case `One` indicate no need to pair. This is needed because zkapp command
+         might be left in pool of half completion. *)
+      type t =
+        { one_or_two : [ `First | `Second | `One ]
+        ; pair_uuid : UUID.Stable.V1.t option
+        }
+      [@@deriving compare, hash, sexp, yojson, equal]
+
+      let to_latest = Fn.id
+    end
+  end]
+end
+
+module Zkapp_command_job = struct
+  (* This identifies a single `Zkapp_command_job.t` *)
+  module UUID = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type t = Job_UUID of int [@@deriving compare, hash, sexp, yojson]
+
+        let to_latest = Fn.id
+      end
+    end]
+  end
+
+  module Spec = struct
+    [%%versioned
+    module Stable = struct
+      [@@@no_toplevel_latest_type]
```

Comment:
I don't know the impact of several `no_top_level_latest_type`. When I was implementing this PR in a draft on an older branch, there's no such restriction.